### DISCUSSION TITLE
Trouble writing tag to mpeg file containing only audio data

### DISCUSSION THREAD

**Discussion by sandersantema** [Serial-ATA/lofty-rs#219] [Q&A] [answered]
EDIT: for anyone else seeing this, this was a bug which was fixed in https://github.com/Serial-ATA/lofty-rs/pull/227

I don't know for certain whether I've run into a bug or if I'm just not using lofty correctly. 

I've got mpeg files containing essentially nothing but audio[1] to which I'm trying to write a new `Id3V2` tag. For the flac files this works fine in my program:
```rust
self.tag.re_map(TagType::VorbisComments);
self.tag.save_to_path(path)?;
```

I've tried the same for the mpeg files, namely `self.tag.re_map(TagType::Id3v2);` but that didn't work.
Nor did directly writing to the file handler:
```rust
let tag: Id3v2Tag = self.tag.into();
let mut mpeg_file = MpegFile::read_from(&mut file_handler, ParseOptions::default())?;
let _ = mpeg_file.set_id3v2(tag);
```

I'd like to know the correct way to write the tag. If the above should already work or the correct way doesn't after testing I'll go ahead and open a proper bug report. If needed I can also provide one of the mpeg mp3 files which fails.

[1] Afaik at least, looking at the file with a hex-editor it starts with zeroes until `000001a0`:
`00ff fbe0 2400 0000 0000 0000 0000 0000`
continues with zeroes until `000005b0`: 
`0000 0000 00ff fbe0 0400 0106 cb82 bab3`
After which the actual data seems to start.

**Comment by Serial-ATA** [comment]
Hello!

The first example when using `TagType::Id3v2` should work. What errors are you getting?

The second example is only setting the tag for the file in memory, you'll need to either do `tag.save_to()` or `mpeg_file.save_to()`, as `MpegFile` is not tied to any file handle by design.

> Afaik at least, looking at the file with a hex-editor it starts with zeroes until

Lofty will skip invalid padding at the start of MPEG files, so that shouldn't be an issue. If the above doesn't work, you can email the file to `serial (AT) [domain on my profile]` (There's a 25MB attachment limit at the moment).

**Reply by sandersantema** [reply]
Oops my post was a bit sloppy, the actual line of code writing the file to disk got lost in all the other stuff I tried. 

The following: 
```rust
let tag: Id3v2Tag = self.tag.into();
let mut mpeg_file = MpegFile::read_from(&mut file_handler, ParseOptions::default())?;
let _ = mpeg_file.set_id3v2(tag);
file_handler.write_all(&self.data)?;
```
Gives the error
```
Error: failed to fill whole buffer
```

And this:
```rust
self.tag.re_map(TagType::Id3v2);
self.tag.save_to_path(path)?;
```
where the `path` is `/home/<...>/test/Fairy Dust EP/5. Set Up System - Music and Noise.mp3`
Gives the error
```
Error: No format could be determined from the provided file
```

The file is small enough but I couldn't get github to upload it (not as an mp3 or jpg, zip), so I've sent you an email.

Thank you very much for the quick response and you/ other contributors for the great library. I don't think there's any out there which supports so many filetypes with a consistent api for all of them, really great!

**Reply by sandersantema** [reply]
Some more variations:

```rust
let tag: Id3v2Tag = self.tag.into();
let mut mpeg_file = MpegFile::read_from(&mut file_handler, ParseOptions::default())?;
let _ = mpeg_file.set_id3v2(tag);
mpeg_file.save_to_path(path)?;
```
Gives: `Error: failed to fill whole buffer`


```rust
let tag: Id3v2Tag = self.tag.into();
tag.save_to(&mut file_handler)?;
file_handler.write_all(&self.data)?;
```
Gives: `Error: Attempted to write a tag to a format that does not support it`

**Reply by Serial-ATA** [reply]
> Error: No format could be determined from the provided file

Ah, yeah this is a bug. I forgot that the invalid padding will only be skipped if there's an ID3v2 tag at the start of the file, otherwise it just expects the file to start with a frame sync.

MP3 isn't that great of a format for easy parsing. The only somewhat reasonable solution I can come up with is to search the first 1024 bytes for a frame sync. Hopefully that won't be too wasteful.

**Reply by sandersantema** [reply]
Are you referring to this line?
https://github.com/Serial-ATA/lofty-rs/blob/307ef146d10b97092daba1325590d2dfacc912a7/src/mpeg/header.rs#L51

Increasing that to 2048 produces the same error but I guess that isn't what you are referring to.

**Reply by Serial-ATA** [reply]
Oh no, sorry. I meant if a format is not determined it will search the first 1024 bytes of the file to find the frame sync. That's not implemented anywhere yet, just an idea of how to go about this.

`REV_FRAME_SEARCH_BOUNDS` is used somewhere completely different.

**Reply by sandersantema** [reply]
I see, if its not too much asked would you happen to know of a workaround using lofty in the meantime? I prefer a workaround in lofty but luckily already found that ffmpeg repairs the header if the copy method is used: `ffmpeg -i input.mp3 -c copy output.mp3`, so if there's no good workaround for lofty I'll just use that for a while :)

**Reply by Serial-ATA** [reply]
Unfortunately, any tag saving will use `Probe::guess_file_type()` internally, so there's no way to workaround it.

**Reply by Serial-ATA** [reply, marked as answer]
@sandersantema This was actually simpler to fix than I thought. Could you test out https://github.com/Serial-ATA/lofty-rs/pull/227 and see if it's able to determine the format now?

**Reply by sandersantema** [reply]
Done, thank you for fixing this so quickly!
